### PR TITLE
OOP: Remove DiagnosticData.Workspace

### PR DESCRIPTION
**Pull request by tmat**
Remove dependency on mutable state. The workspace the diagnostic data originated from is available thru `DiagnosticsUpdatedArgs` and the `DiagnosticData` do not need the reference.

### CODE REVIEW

**Comment by tmat** [comment]
@heejaechang PTAL

**Comment by tmat** [comment]
@dotnet/roslyn-ide

File: src/Features/Core/Portable/Diagnostics/AnalyzerHelper.cs [hunk 1]
```diff
@@ -211,11 +211,10 @@ private static VersionStamp GetAnalyzerVersion(string path)
 
         public static DiagnosticData CreateAnalyzerLoadFailureDiagnostic(string fullPath, AnalyzerLoadFailureEventArgs e)
         {
-            return CreateAnalyzerLoadFailureDiagnostic(null, null, null, fullPath, e);
+            return CreateAnalyzerLoadFailureDiagnostic(null, null, fullPath, e);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Diagnostics/AnalyzerHelper.cs]: named params would help here.

File: src/VisualStudio/Core/Def/Implementation/TableDataSource/TableItem.cs [hunk 2]
```diff
@@ -74,41 +81,34 @@ public TableItem(IEnumerable<TableItem<T>> items)
...
 
+        // item must be either one of diagnostic data and todo item
         public DocumentId PrimaryDocumentId
-        {
-            get
-            {
-                return Extensions.GetDocumentId(Primary);
-            }
-        }
+            => (Primary is DiagnosticData diagnostic) ? diagnostic.DocumentId : ((TodoItem)(object)Primary).DocumentId;
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/TableDataSource/TableItem.cs]: man do i dislike this.

File: src/VisualStudio/Core/Def/Implementation/TableDataSource/VisualStudioBaseTodoListTable.cs [hunk 3]
```diff
@@ -303,19 +303,38 @@ private LinePosition GetLineColumn(TodoItem item)
...
+                    {
+                        line = trackingLinePosition.Line;
+                        column = trackingLinePosition.Character;
+                    }
+                    else
                     {
-                        return TryNavigateTo(item.Workspace, item.DocumentId, trackingLinePosition.Line, trackingLinePosition.Character, previewTab);
+                        line = primary.OriginalLine;
+                        column = primary.OriginalColumn;
                     }
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/TableDataSource/VisualStudioBaseTodoListTable.cs]: this logic loos very duplicative with what's two files above.  No way to share more?

File: src/VisualStudio/Core/Def/Implementation/TaskList/ExternalErrorDiagnosticUpdateSource.cs [hunk 4]
```diff
@@ -689,82 +689,38 @@ private class DiagnosticDataComparer : IEqualityComparer<DiagnosticData>
...
+                    item1.ProjectId != item2.ProjectId ||
+                    item1.Severity != item2.Severity ||
+                    item1.Message != item2.Message ||
+                    (item1.DataLocation?.MappedStartLine ?? 0) != (item2.DataLocation?.MappedStartLine ?? 0) ||
+                    (item1.DataLocation?.MappedStartColumn ?? 0) != (item2.DataLocation?.MappedStartColumn ?? 0) ||
+                    (item1.DataLocation?.OriginalStartLine ?? 0) != (item2.DataLocation?.OriginalStartLine ?? 0) ||
+                    (item1.DataLocation?.OriginalStartColumn ?? 0) != (item2.DataLocation?.OriginalStartColumn ?? 0))
                 {
                     return false;
                 }
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/TaskList/ExternalErrorDiagnosticUpdateSource.cs]: this seems strange...  why not just do stuff like `Equals(item1.DocumentId, item2.DocumentId)`?

File: src/Workspaces/Core/Portable/Diagnostics/InternalDiagnosticsOptions.cs [hunk 5]
```diff
@@ -8,9 +8,6 @@ internal static class InternalDiagnosticsOptions
     {
         private const string LocalRegistryPath = @"Roslyn\Internal\Diagnostics\";
 
-        public static readonly Option<bool> BlueSquiggleForBuildDiagnostic = new Option<bool>(nameof(InternalDiagnosticsOptions), "Blue Squiggle For Build Diagnostic", defaultValue: false,
-            storageLocations: new LocalUserProfileStorageLocation(LocalRegistryPath + "Blue Squiggle For Build Diagnostic"));
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Diagnostics/InternalDiagnosticsOptions.cs]: pure question: why was this removed?  I have no problem with it going.  I'm just trying to understand the reasoning/discussion around it.

**Comment by CyrusNajmabadi** [comment]
This is great.  Nothing struck me as bad here.  Though i didn't look closely at some parts of hte logic that seem a little on hte complex side.  

One concern i have (which may be better than how things worked before) is how resilient we are to the diagnostic data being stale wrt the current state of the workspace.  Is the code the interacts between teh two suitably safe toward drift between them?

File: src/VisualStudio/Core/Def/Implementation/TableDataSource/TableItem.cs (hunk 2 above)
Comment (tmat) [reply src/VisualStudio/Core/Def/Implementation/TableDataSource/TableItem.cs]: Me too, will follow up with more refactoring to clean this up.

File: src/VisualStudio/Core/Def/Implementation/TableDataSource/VisualStudioBaseTodoListTable.cs (hunk 3 above)
Comment (tmat) [reply src/VisualStudio/Core/Def/Implementation/TableDataSource/VisualStudioBaseTodoListTable.cs]: I think we can, but needs a bit more refactoring. I'll follow up.

---
In reply to: [287435304](https://github.com/dotnet/roslyn/pull/35902#discussion_r287435304) [](ancestors = 287435304)

File: src/Workspaces/Core/Portable/Diagnostics/InternalDiagnosticsOptions.cs (hunk 5 above)
Comment (tmat) [reply src/Workspaces/Core/Portable/Diagnostics/InternalDiagnosticsOptions.cs]: It's just not needed anymore.

---
In reply to: [287436212](https://github.com/dotnet/roslyn/pull/35902#discussion_r287436212) [](ancestors = 287436212)

File: src/VisualStudio/Core/Def/Implementation/TaskList/ExternalErrorDiagnosticUpdateSource.cs (hunk 4 above)
Comment (tmat) [reply src/VisualStudio/Core/Def/Implementation/TaskList/ExternalErrorDiagnosticUpdateSource.cs]: @heejaechang ?

Comment (heejaechang) [reply src/VisualStudio/Core/Def/Implementation/TaskList/ExternalErrorDiagnosticUpdateSource.cs]: the code looks like checking whether 2 are both null or not null before going in to even check equality. not sure why it is added. though. probably some perf optimization due to some random trace.

File: src/Workspaces/Core/Portable/Diagnostics/InternalDiagnosticsOptions.cs (hunk 5 above)
Comment (heejaechang) [reply src/Workspaces/Core/Portable/Diagnostics/InternalDiagnosticsOptions.cs]: it was added to test user experience long time ago and to test deduplication and no longer used. and we made decision not to show different color for build errors.

File: src/VisualStudio/Core/Def/Implementation/TableDataSource/TableItem.cs (hunk 2 above)
Comment (heejaechang) [reply src/VisualStudio/Core/Def/Implementation/TableDataSource/TableItem.cs]: ya. you can make TableItem to hold onto document and projectId directly rather than get it through item or pass in lambda to get those or make diagnostic/todoItem to implement interface to get document/project id.

File: src/VisualStudio/Core/Def/Implementation/TableDataSource/VisualStudioBaseTodoListTable.cs (hunk 3 above)
Comment (tmat) [reply src/VisualStudio/Core/Def/Implementation/TableDataSource/VisualStudioBaseTodoListTable.cs]: Done.

File: src/Features/Core/Portable/Diagnostics/AnalyzerHelper.cs (hunk 1 above)
Comment (tmat) [reply src/Features/Core/Portable/Diagnostics/AnalyzerHelper.cs]: Done

File: src/VisualStudio/Core/Def/Implementation/TableDataSource/TableItem.cs (hunk 2 above)
Comment (tmat) [reply src/VisualStudio/Core/Def/Implementation/TableDataSource/TableItem.cs]: Fixed.

File: src/Workspaces/Core/Portable/Diagnostics/InternalDiagnosticsOptions.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Diagnostics/InternalDiagnosticsOptions.cs]: Thanks for the explanation.  It's very appreciated!

File: src/Features/Core/Portable/Diagnostics/AnalyzerHelper.cs [hunk 6]
```diff
@@ -211,11 +211,10 @@ private static VersionStamp GetAnalyzerVersion(string path)
 
         public static DiagnosticData CreateAnalyzerLoadFailureDiagnostic(string fullPath, AnalyzerLoadFailureEventArgs e)
         {
-            return CreateAnalyzerLoadFailureDiagnostic(null, null, null, fullPath, e);
+            return CreateAnalyzerLoadFailureDiagnostic(projectId: null, language: null, fullPath, e);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Diagnostics/AnalyzerHelper.cs]: thanks

File: src/VisualStudio/Core/Def/Implementation/TableDataSource/AbstractTableDataSource.cs [hunk 7]
```diff
@@ -103,9 +102,12 @@ public void Shutdown()
             }
         }
 
-        public abstract ImmutableArray<TableItem<TData>> Deduplicate(IEnumerable<IList<TableItem<TData>>> duplicatedGroups);
-        public abstract ITrackingPoint CreateTrackingPoint(TData data, ITextSnapshot snapshot);
-        public abstract AbstractTableEntriesSnapshot<TData> CreateSnapshot(AbstractTableEntriesSource<TData> source, int version, ImmutableArray<TableItem<TData>> items, ImmutableArray<ITrackingPoint> trackingPoints);
+        public ImmutableArray<TItem> Deduplicate(IEnumerable<IList<TItem>> groupedItems)
+            => groupedItems.MergeDuplicatesOrderedBy(Order);
+
+        public abstract IEnumerable<TItem> Order(IEnumerable<TItem> groupedItems);
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/TableDataSource/AbstractTableDataSource.cs]: maybe doc this a bit... it's not clear to me what'd going on.

File: src/VisualStudio/Core/Def/Implementation/TableDataSource/AbstractTableEntriesSnapshot.cs [hunk 8]
```diff
@@ -165,44 +141,70 @@ protected LinePosition GetTrackingLineColumn(Workspace workspace, DocumentId doc
...
+            if (document == null)
+            {
+                return false;
+            }
+
+            LinePosition position;
+            LinePosition trackingLinePosition;
+
+            if (workspace.IsDocumentOpen(documentId) &&
+                (trackingLinePosition = GetTrackingLineColumn(document, index)) != LinePosition.Zero)
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/TableDataSource/AbstractTableEntriesSnapshot.cs]: i'm not sure i understand what's going on here.  can you doc?

File: src/VisualStudio/Core/Def/Implementation/TableDataSource/DiagnosticTableItem.cs [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+            : base(workspace, cache)
+        {
+            Contract.ThrowIfNull(data);
+            Data = data;
+        }
+
+        public override TableItem WithCache(SharedInfoCache cache)
+            => new DiagnosticTableItem(Workspace, cache, Data);
+
+        public override DocumentId PrimaryDocumentId
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/TableDataSource/DiagnosticTableItem.cs]: what does primary here mean?  is this in reference to files that are linked?

File: src/VisualStudio/Core/Def/Implementation/TableDataSource/DiagnosticTableItem.cs [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+                Hash.Combine(diagnostic.DataLocation.OriginalStartLine,
+                Hash.Combine(diagnostic.DataLocation.OriginalEndColumn,
+                Hash.Combine(diagnostic.DataLocation.OriginalEndLine,
+                Hash.Combine(diagnostic.DataLocation.OriginalFilePath,
+                Hash.Combine(diagnostic.IsSuppressed,
+                Hash.Combine(diagnostic.Id.GetHashCode(), diagnostic.Message.GetHashCode())))))));
+        }
+
+        public override LinePosition GetOriginalPosition()
+            => new LinePosition(Data.DataLocation?.OriginalStartLine ?? 0, Data.DataLocation?.OriginalStartColumn ?? 0);
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/TableDataSource/DiagnosticTableItem.cs]: this looks identical to GetTrackingPosition.  Is there a difference?  do we need both?

File: src/VisualStudio/Core/Def/Implementation/TableDataSource/Extensions.cs [hunk 11]
```diff
@@ -31,38 +26,47 @@ internal static class Extensions
...
+            var builder = ArrayBuilder<TItem>.GetInstance();
+            foreach (var item in orderer(groupedItems.Select(Deduplicate)))
             {
                 builder.Add(item);
             }
 
             return builder.ToImmutableAndFree();
         }
 
-        private static TableItem<T> Deduplicate<T>(this IList<TableItem<T>> duplicatedItems)
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/TableDataSource/Extensions.cs]: pausing here.

File: src/VisualStudio/Core/Def/Implementation/TableDataSource/DiagnosticTableItem.cs (hunk 10 above)
Comment (tmat) [reply src/VisualStudio/Core/Def/Implementation/TableDataSource/DiagnosticTableItem.cs]: Removed.

File: src/VisualStudio/Core/Def/Implementation/TableDataSource/DiagnosticTableItem.cs (hunk 9 above)
Comment (tmat) [reply src/VisualStudio/Core/Def/Implementation/TableDataSource/DiagnosticTableItem.cs]: Renamed.
